### PR TITLE
Refactor access token to make it extensible

### PR DESCRIPTION
**Pull request by raymondfeng**
1. Make it possible to reuse getAccessTokenId
2. Introduce a flag to control if oAuth2 bearer token should be base64
encoded

#### Related issues

- None

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: common/models/access-token.js [hunk 1]
```diff
@@ -78,6 +78,77 @@ module.exports = function(AccessToken) {
     });
   });
 
+  AccessToken.getAccessTokenId = function(req, options) {
```
Comment (ritch) [inline common/models/access-token.js]: Public api now? Needs doc before we merge.

Comment (ritch) [inline common/models/access-token.js]: Since the method moved its impossible to do a diff...? Is this unchanged otherwise?

File: server/middleware/token.js [hunk 2]
```diff
@@ -88,6 +88,8 @@ function escapeRegExp(str) {
  * @property {Boolean} [overwriteExistingToken] only has effect in combination with `enableDoublecheck`. If truthy, will allow to overwrite an existing accessToken.
  * @property {Function|String} [model] AccessToken model name or class to use.
  * @property {String} [currentUserLiteral] String literal for the current user.
+ * @property {Boolean} [bearerTokenBase64Encoded] oAuth2 is not clear if bearer token in Authorization header should be base64 encoded.
```
Comment (ritch) [inline server/middleware/token.js]:
```
oAuth2 is not clear if bearer token in Authorization header should be base64 encoded.
If set, we'll decode the token first. Default to `true` for backward compatibility.
```

I would reword this a little bit:

```
Defaults to `true`. For `Bearer` token based `Authorization` headers, decode the value from `Base64`. If set to `false` the decoding will be skipped and the token id will be the raw value parsed from the header.
```

File: common/models/access-token.js (hunk 1 above)
Comment (raymondfeng) [inline common/models/access-token.js]: Yes. It's moved out as is. The jsdoc is added now.

**Comment by raymondfeng** [comment]
@ritch PTAL.

File: test/access-token.test.js [hunk 3]
```diff
@@ -462,6 +473,11 @@ describe('loopback.token(options)', function() {
 describe('AccessToken', function() {
   beforeEach(createTestingToken);
 
+  it('should have getAccessTokenId method', function() {
+    assert(typeof Token.getAccessTokenId === 'function');
+    assert(typeof Token.findForRequest === 'function');
+  });
```
Comment (ebarault) [inline test/access-token.test.js]: - there should be one assertion per test (test on `.findForRequest()` is unrelated)
- new tests should use expect, not assert
- new test descriptions do not mention 'should': `it has getAccessTokenId method`
- test on `.getAccessTokenId()` should actually verify that the method can be customized by the user (main purpose on the PR)
- while at it, test on `. findForRequest()` should do likewise (as it's public api)

**Comment by raymondfeng** [comment]
@ebarault Thanks for the feedback. PTAL.

File: common/models/access-token.js [hunk 4]
```diff
@@ -79,6 +79,119 @@ module.exports = function(AccessToken) {
...
+    }
+    return null;
+  };
+
+  /**
+   * Validate the access token
+   * @param {String} id Access token
+   * @callback {Fuction} cb Callback function
+   */
+  AccessToken.validate = function(id, cb) {
```
Comment (ebarault) [inline common/models/access-token.js]: - the signature of this new api is inconsistent with `.prototype.validate()` which returns a **Boolean**, i suggest returning a boolean, as the token is already provided as param.
- please add also the @callback signature in the jsdoc.
```
// as in
  @callback (Function) callback
  @param {Error} err
  @param {isValid} is the token valid
```
- you also have a typo (see: _Fuction_)

File: common/models/access-token.js [hunk 5]
```diff
@@ -79,6 +79,119 @@ module.exports = function(AccessToken) {
...
+    this.findById(id, function(err, token) {
+      if (err) {
+        cb(err);
+      } else if (token) {
+        token.validate(function(err, isValid) {
+          if (err) {
+            cb(err);
+          } else if (isValid) {
+            cb(null, token);
```
Comment (ebarault) [inline common/models/access-token.js]: `cb(null, true);`

File: common/models/access-token.js [hunk 6]
```diff
@@ -87,40 +200,18 @@ module.exports = function(AccessToken) {
...
-              e.status = e.statusCode = 401;
-              e.code = 'INVALID_TOKEN';
-              cb(e);
-            }
-          });
-        } else {
-          cb();
-        }
-      });
+      this.validate(id, cb);
```
Comment (ebarault) [inline common/models/access-token.js]:
```
this.validate(id, function(err, isValid) {
  if (err) cb(err);
  cb(null, isValid);
});
```

File: test/access-token.test.js [hunk 7]
```diff
@@ -172,6 +173,16 @@ describe('loopback.token(options)', function() {
       .end(done);
   });
 
+  it('should populate req.token from an authorization header with bearer token', function(done) {
```
Comment (ebarault) [inline test/access-token.test.js]: ditto

File: test/access-token.test.js [hunk 8]
```diff
@@ -162,7 +162,8 @@ describe('loopback.token(options)', function() {
     });
   });
 
-  it('should populate req.token from an authorization header with bearer token', function(done) {
+  it('should populate req.token from an authorization header with bearer token with base64',
```
Comment (ebarault) [inline test/access-token.test.js]: while at it could you remove the `should`?

File: test/access-token.test.js [hunk 9]
```diff
@@ -525,6 +544,54 @@ describe('AccessToken', function() {
...
+        if (err) return done(err);
+
+        expect(token.id).to.eql(expectedTokenId);
+        expect(called).to.be.true();
+
+        done();
+      });
+    });
+
     function mockRequest(opts) {
```
Comment (ebarault) [inline test/access-token.test.js]: this should do

**Comment by raymondfeng** [comment]
@ebarault I made the further changes per your request:

1. Rename `AccessToken.validate` to `AccessToken.resolve`. Please note that `resolve` takes `id` and returns the full access token object.

2. Fix the jsdoc

3. Remove `should` for all tests

**Comment by raymondfeng** [comment]
We cannot change `findForRequest` as it's already there.

I can rename as follows:

1. getAccessTokenId --> getIdForRequest
2. resolve (as-is, resolve already hints by id)
